src/mpris-backend/mod.rs

```diff
@@ -0,0 +1,209 @@
+
+pub enum PlayerctlAction {
+	PlayPause,
+	Play,
+	Pause,
+	Stop,
+	Next,
+	Prev,
+	Shuffle,
+}
+
+use super::config::user::ServerConfig;
+
+#[derive(Clone, Debug)]
+pub enum PlayerctlDeviceRaw {
+	None,
+	All,
+	Some(String),
+}
+
+pub enum PlayerctlDevice {
+	All(Vec<Player>),
+	Some(Player),
+}
+
+pub struct Playerctl {
+	player: PlayerctlDevice,
+	action: PlayerctlAction,
+	pub icon: Option<String>,
+	pub label: Option<String>,
+	fmt_str: Option<String>,
+}
+
+impl Playerctl {
+	pub fn new(
+		action: PlayerctlAction,
+		config: Arc<ServerConfig>,
+	) -> Result<Playerctl, Box<dyn Error>> {
+		let playerfinder = PlayerFinder::new()?;
+		let player = get_player();
+		let player = match player {
+			PlayerctlDeviceRaw::None => PlayerctlDevice::Some(playerfinder.find_active()?),
+			PlayerctlDeviceRaw::Some(name) => {
+				PlayerctlDevice::Some(playerfinder.find_by_name(name.as_str())?)
+			}
+			PlayerctlDeviceRaw::All => PlayerctlDevice::All(playerfinder.find_all()?),
+		};
+		let fmt_str = config.playerctl_format.clone();
+		Ok(Self {
+			player,
+			action,
+			icon: None,
+			label: None,
+			fmt_str,
+		})
+	}
+	pub fn run(&mut self) -> Result<(), Box<dyn Error>> {
+		use PlaybackStatus::*;
+		use PlayerctlAction::*;
+		let run_single = |player: &Player| -> Result<&str, Box<dyn Error>> {
+			let out = match self.action {
+				PlayPause => match player.get_playback_status()? {
+					Playing => {
+						player.pause()?;
+						"pause-large-symbolic"
+					}
+					Paused | Stopped => {
+						player.play()?;
+						"play-large-symbolic"
+					}
+				},
+				Shuffle => {
+					let shuffle = player.get_shuffle()?;
+					player.set_shuffle(!shuffle)?;
+					if shuffle {
+						"playlist-consecutive-symbolic"
+					} else {
+						"playlist-shuffle-symbolic"
+					}
+				}
+				Play => {
+					player.play()?;
+					"play-large-symbolic"
+				}
+				Pause => {
+					player.pause()?;
+					"pause-large-symbolic"
+				}
+				Stop => {
+					player.stop()?;
+					"stop-large-symbolic"
+				}
+				Next => {
+					player.next()?;
+					"media-seek-forward-symbolic"
+				}
+				Prev => {
+					player.previous()?;
+					"media-seek-backward-symbolic"
+				}
+			};
+			Ok(out)
+		};
+		let mut metadata = Err("some errro");
+		let icon = match &self.player {
+			PlayerctlDevice::Some(player) => {
+				metadata = player.get_metadata().or_else(|_| Err(""));
+				run_single(player)?
+			}
+			PlayerctlDevice::All(players) => {
+				let mut icon = Err("couldn't change any players!");
+				for player in players {
+					let icon_new = run_single(player);
+					if let Ok(icon_new) = icon_new {
+						if icon.is_err() {
+							icon = Ok(icon_new);
+						}
+					};
+					if let Err(_) = metadata {
+						metadata = player.get_metadata().or_else(|_| Err(""));
+					}
+				}
+				icon?
+			}
+		};
+
+		self.icon = Some(icon.to_string());
+		let label = if let Ok(metadata) = metadata {
+			Some(self.fmt_string(metadata))
+		} else {
+			None
+		};
```

Comment:
The player name or something similar should be displayed instead when running next/prev due to the wrong title being displayed. Ex:

Playing song "A" and running the next command displays the "A" song while it should display "B".

Not really sure if displaying B is possible without delaying and waiting for the next metadata to come in, so an easy fix would be to display the player name :)